### PR TITLE
Update deployment script to accept python3 WMAgent service

### LINKED ISSUE #10230: Create deployment scripts for python3 WMAgent
**Impact of the new feature**
WMAgent

**Is your feature request related to a problem? Please describe.**
Yes, python2 is no longer supported ;)

**Describe the solution you'd like**
Once WMAgent python3 specs have been created, we will need to make deployment changes - or create - for the python3 WMAgent system.

**Describe alternatives you've considered**
none

**Additional context**
none

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #10230 

#### Status
tested

#### Description
Add a new argument to the deploy-wmagent script such that we can specify which RPM we would like to deploy:
* wmagent (python2)
* wmagentpy3 (python3 stack)

NOTE: when deploying a python3 agent, we need to update the `env.sh` script to point to:
```
export install=/data/srv/wmagent/current/install/wmagentpy3
export config=/data/srv/wmagent/current/config/wmagentpy3
```
instead of:
```
export install=/data/srv/wmagent/current/install/wmagent
export config=/data/srv/wmagent/current/config/wmagent
```

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
none

#### External dependencies / deployment changes
Deployment changes: https://github.com/dmwm/deployment/pull/1034

### CODE REVIEW

**Comment by amaltaro** [comment]
With the proposed changes, WMAgent and all its dependencies can be installed. Unfortunately things fail to get initialized and the final agent deployment setup cannot be completed. This is the error message from the logs:
```
*** Initializing the agent ***
Initialising Agent...
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.6.pre7/sw/slc7_amd64_gcc630/cms/wmagentpy3/1.4.6.pre7/bin/wmagent-mod-config", line 14, in <module>
    standard_library.install_aliases()
  File "/data/srv/wmagent/v1.4.6.pre7/sw/slc7_amd64_gcc630/external/py3-future/0.18.2/lib/python3.8/site-packages/future/standard_library/__init__.py", line 453, in install_aliases
    __import__(newmodname)
ImportError: No module named reprlib
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.6.pre7/sw/slc7_amd64_gcc630/cms/wmagentpy3/1.4.6.pre7/bin/wmcore-db-init", line 19, in <module>
    from WMCore.Configuration import loadConfigurationFile
ImportError: No module named WMCore.Configuration
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.6.pre7/sw/slc7_amd64_gcc630/cms/wmagentpy3/1.4.6.pre7/bin/wmagent-couchapp-init", line 9, in <module>
    standard_library.install_aliases()
  File "/data/srv/wmagent/v1.4.6.pre7/sw/slc7_amd64_gcc630/external/py3-future/0.18.2/lib/python3.8/site-packages/future/standard_library/__init__.py", line 453, in install_aliases
    __import__(newmodname)
ImportError: No module named reprlib
Done!
```

and we are ending up with a mix of python2 and python3 libraries (we haven't yet migrated away of some python2-only libraries), examples are:
* couchapp (there is no support to python3): https://github.com/cms-sw/cmsdist/blob/comp_gcc630/couchapp.spec
* we might have to build condor python bindings under python3 (even better if we manage to build it via pip): https://github.com/cms-sw/cmsdist/blob/comp_gcc630/condor.spec
* need a python3 package for pystack: https://github.com/cms-sw/cmsdist/blob/comp_gcc630/pystack.spec 
* need a python3 package for cmsmonitoring: https://github.com/cms-sw/cmsdist/blob/comp_gcc630/cmsmonitoring.spec
* and need a python3 package for https://github.com/cms-sw/cmsdist/blob/comp_gcc630/dbs3-client.spec

I will make some new issues soon.

**Comment by amaltaro** [comment]
In terms of deployment changes, I think this covers what is needed. The whole wmagentpy3 deployment is not fully working yet because there are many other issues with RPMs and their dependencies.

File: deploy/deploy-wmagent.sh [hunk 1]
```diff
@@ -285,12 +293,15 @@ echo -e "\n*** Applying agent patches ***"
 if [ "x$PATCHES" != "x" ]; then
   cd $CURRENT_DIR
   for pr in $PATCHES; do
-    wget -nv https://github.com/dmwm/WMCore/pull/$pr.patch -O - | patch -d apps/wmagent/lib/python2*/site-packages/ -p 3
+    wget -nv https://github.com/dmwm/WMCore/pull/$pr.patch -O - | patch -d apps/${RPM_NAME}/lib/python2*/site-packages/ -p 3
```
Comment (todor-ivanov) [inline deploy/deploy-wmagent.sh:296]: I am pretty sure you have already tested it, but just my curiosity: Why hard coded path leading to `python2` libraries even though here we provide the option to deploy a python3 version of the rpm package

File: deploy/deploy-wmagent.sh [hunk 2]
```diff
@@ -269,14 +271,20 @@ cd $BASE_DIR/deployment-$DEPLOY_TAG
...
 for step in prep sw post; do
   echo -e "\n*** Deploying WMAgent: running $step step ***"
-  ./Deploy -R wmagent@$WMA_TAG -s $step -A $WMA_ARCH -r $REPO -t v$WMA_TAG $DEPLOY_DIR wmagent
+  ./Deploy -R ${RPM_NAME}@$WMA_TAG -s $step -A $WMA_ARCH -r $REPO -t v$WMA_TAG $DEPLOY_DIR ${RPM_NAME}
 done
 set +e
 
+echo -e "\n*** Creating wmagent symlink ***"
+cd $CURRENT_DIR
+ln -s ../sw/${WMA_ARCH}/cms/${RPM_NAME}/${WMA_TAG} apps/wmagent
```
Comment (todor-ivanov) [inline deploy/deploy-wmagent.sh:280]: Here you seem to create the link name always as `apps/wmagent`, which to me seems logical, but down bellow you refer to it as `apps/${RPM_NAME}` (line 296 in the current commit)

**Review by todor-ivanov — COMMENTED** [review]
@amaltaro I just made two comments inline. I hope you have already tested the current deployment script with the python3 version of the rpm package and it is working, but those lines that I am pointing to seem confusing. Could you please take a look.

File: deploy/deploy-wmagent.sh (hunk 2 above)
Comment (amaltaro) [reply deploy/deploy-wmagent.sh:280]: I had to create this symlink because some of those bin/* scripts look for files under `apps/wmagent`, which does not exist when we deploy `wmagentpy3`. 
The Deploy script automatically creates a symlink according to the package name. So, for python2, it won't make any difference since the symlink will be already there. For python3, we will have both symlinks pointing to the same directory structure.

In short, I could simply use `apps/wmagent` below as well, but I tried to keep it as consistent as possible.

File: deploy/deploy-wmagent.sh (hunk 1 above)
Comment (amaltaro) [reply deploy/deploy-wmagent.sh:296]: Good catch! This needs to be fixed.

**Comment by amaltaro** [comment]
Thanks for the review, Todor. I fixed one issue that you spotted.
Just wanted to mention that, ideally, I'd like to use this new RPM/package name (wmagentpy3) only for a transition period. Once WMCore/WMAgent is fully validated and functional in python3, we could likely rename it back to `wmagent` and drop the python2-based agent.

**Review by todor-ivanov — APPROVED** [review]
looks good to me

**Comment by todor-ivanov** [comment]
Thanks @amaltaro. the temporary name sounds logical to me too.

**Comment by amaltaro** [comment]
Thanks Todor.
